azurelinuxagent/common/cgroupconfigurator.py

```diff
@@ -248,37 +237,52 @@ def check_processes_in_agent_cgroup(self):
-                if current == 0:
-                    unexpected.append(process)
-                    if len(unexpected) >= 5:  # collect just a small sample
-                        break
-            if unexpected:
-                raise UnexpectedProcessesInCGroupException(unexpected=self._format_processes(unexpected))
+            def log_message(message):
+                # Report only a small sample of errors
+                if message != self._check_processes_in_agent_cgroup_last_error and self._check_processes_in_agent_cgroup_error_count < 5:
+                    self._check_processes_in_agent_cgroup_error_count += 1
+                    self._check_processes_in_agent_cgroup_last_error = message
+                    logger.info(message)
+                    add_event(op=WALAEventOperation.CGroupsDebug, message=message)
+
+            try:
+                daemon = os.getppid()
+                extension_handler = os.getpid()
+                agent_commands = set()
+                agent_commands.update(shellutil.get_running_commands())
+                systemd_run_commands = set()
+                systemd_run_commands.update(self._cgroups_api.get_systemd_run_commands())
+                agent_cgroup = CGroupsApi.get_processes_in_cgroup(self._agent_cpu_cgroup_path)
+                # get the running commands again in case new commands were started while we were fetching the processes in the cgroup;
+                agent_commands.update(shellutil.get_running_commands())
+                systemd_run_commands.update(self._cgroups_api.get_systemd_run_commands())
+
+                unexpected = []
+                for process in agent_cgroup:
+                    # Note that the agent uses systemd-run to start extensions; systemd-run belongs to the agent cgroup, though the extensions don't
+                    if process in (daemon, extension_handler) or process in systemd_run_commands:
+                        continue
+                    # check if the process is a command started by the agent or a descendant of one of those commands
+                    current = process
+                    while current != 0 and current not in agent_commands:
+                        current = self._get_parent(current)
+                    if current == 0:
+                        unexpected.append(process)
+                        if len(unexpected) >= 5:  # collect just a small sample
+                            break
+                if unexpected:
+                    unexpected = self._format_processes(unexpected)
+                    unexpected.sort()
+                    log_message("The agent's cgroup includes unexpected processes; disabling CPU enforcement. Unexpected: {0}".format(unexpected))
+                    self.disable()
+                    return False
+            except Exception as exception:
+                log_message("Failed to check the processes in the agent's cgroup: {0}".format(ustr(exception)))
```

Comment:
Shouldn't we return False here too?

Reply:
At this point the return value is not being used. I am considering removing the return value altogether and also disable cgroups if I can't perform the check, but I need to think a little more on that. Will be coming on a new PR

Reply:
Fair enough